internal/k8s/controller.go

```diff
@@ -871,40 +871,52 @@ func (lbc *LoadBalancerController) syncVirtualServer(task task) {
 		return
 	}
 
+	var vs *conf_v1.VirtualServer
+
+	vslist := lbc.getVirtualServers()
+	for _, v := range vslist {
+		vskey := fmt.Sprintf("%s/%s", v.Namespace, v.Name)
+		if vskey == key {
+			vs = v
+			break
+		}
+	}
+	vsrlist := findVirtualServerRoutesForVirtualServer(vs, lbc.getVirtualServerRoutes())
+
 	if !vsExists {
 		glog.V(2).Infof("Deleting VirtualServer: %v\n", key)
 
 		err := lbc.configurator.DeleteVirtualServer(key)
-		// TO-DO: emit events for referenced VirtualServerRoutes
 		if err != nil {
 			glog.Errorf("Error when deleting configuration for %v: %v", key, err)
+		} else {
+			reason := "NoVirtualServerFound"
+			for _, vsr := range vsrlist {
+				msg := fmt.Sprintf("No VirtualServer references VirtualServerRoute %v/%v", vsr.Namespace, vsr.Name)
+				lbc.recorder.Eventf(vsr, api_v1.EventTypeWarning, reason, msg)
+				lbc.statusUpdater.UpdateVirtualServerRouteStatus(vsr, conf_v1.StateInvalid, reason, msg)
+			}
 		}
+
 		return
 	}
 
 	glog.V(2).Infof("Adding or Updating VirtualServer: %v\n", key)
-
-	vs := obj.(*conf_v1.VirtualServer)
+	vs = obj.(*conf_v1.VirtualServer)
 
 	validationErr := validation.ValidateVirtualServer(vs, lbc.isNginxPlus)
 	if validationErr != nil {
 		err := lbc.configurator.DeleteVirtualServer(key)
 		if err != nil {
 			glog.Errorf("Error when deleting configuration for %v: %v", key, err)
 		}
-		reason := "Rejected"
-		msg := fmt.Sprintf("VirtualServer %v is invalid and was rejected: %v", key, validationErr)
-
-		lbc.recorder.Eventf(vs, api_v1.EventTypeWarning, reason, msg)
-		if lbc.reportVsVsrStatusEnabled() {
-			err = lbc.statusUpdater.UpdateVirtualServerStatus(vs, conf_v1.StateInvalid, reason, msg)
-
-			if err != nil {
-				glog.Errorf("Error when updating the status for VirtualServer %v/%v: %v", vs.Namespace, vs.Name, err)
-			}
+		reason := "NoVirtualServerFound"
+		for _, vsr := range vsrlist {
+			msg := fmt.Sprintf("No VirtualServer references VirtualServerRoute %v/%v", vsr.Namespace, vsr.Name)
+			lbc.recorder.Eventf(vsr, api_v1.EventTypeWarning, reason, msg)
+			lbc.statusUpdater.UpdateVirtualServerRouteStatus(vsr, conf_v1.StateInvalid, reason, msg)
```

Comment:
`UpdateVirtualServerRouteStatus` returns an error. This should be checked?

* If so, this applies in the delete VirtualServer case too.